File: bubble.ts
// I'm creating a function to perform bubble sort
export function bubble_sort(arr: number[]) {
    for (let i = 0; i < arr.length; i++) {
        // this flag variable is to optimize the code 
        // in the best case the time complexity is o(n)
        // while the worst case is o(n^2)
        var flag = 0
        for (let j = 0; j < arr.length - 1 - i; j++) {
            if (arr[j] > arr[j + 1]) {
                let temp = arr[j]
                arr[j] = arr[j + 1]
                arr[j + 1] = temp
                flag = 1
            }
        }
        if (flag == 0) {
            return "array is sorted " + arr;
            break;
        }
    }
}

// This is to test the function
// let arr = [9, 3, 7, 4, 69, 420, 42];
// let sorted_array = bubble_sort(arr);
// console.log(sorted_array)

File: bubble.dart
void main() {
  List<int> myArray = [64, 34, 25, 12, 22, 11, 90];
  List<int> sortedArray = bubbleSort(myArray);
  
  print(sortedArray);
}

List<int> bubbleSort(List<int> arr) {
  for (int i = 0; i < arr.length; i++) {
    // This flag variable is to optimize the code 
    // In the best case, the time complexity is O(n)
    // While the worst case is O(n^2)
    int flag = 0;
    for (int j = 0; j < arr.length - 1 - i; j++) {
      if (arr[j] > arr[j + 1]) {
        int temp = arr[j];
        arr[j] = arr[j + 1];
        arr[j + 1] = temp;
        flag = 1;
      }
    }
    if (flag == 0) {
      return arr;
    }
  }
  return arr;
}
